onnxscript/evaluator.py

```diff
@@ -0,0 +1,310 @@
+
+
+def ort_to_os_value(v):
+    """Converts an ORT encoding of an ONNX value into the encoding used by onnxscript."""
+    if isinstance(v, np.ndarray):
+        return tensor.Tensor(v)
+    if isinstance(v, list):
+        return v
+    if v is None:
+        raise TypeError("Dynamic optional values not yet supported.")
+    raise TypeError(f"Unexpected ORT value type {type(v)}.")
+
+
+def call_ort(schema, args, kwargs, implicit_args=None):
+    implicit_args = implicit_args or {}
+    # Convert input values to ORT representation-type:
+    args = [os_to_ort_value(x) for x in args]
+    implicit_args = {k: os_to_ort_value(v) for k, v in implicit_args.items()}
+
+    # Construct ONNX model with a single op call:
+    inputs = [_rename_io("input", i, arg) for i, arg in enumerate(args)]
+
+    num_outputs = compute_num_outputs(schema, *args, **kwargs)
+    outputs = [f"output{str(i)}" for i in range(num_outputs)]
+
+    node = onnx.helper.make_node(schema.name, inputs, outputs, domain=schema.domain, **kwargs)
+    input_value_infos = utils.values_to_value_infos(zip(inputs, args))
+    implicit_value_infos = utils.values_to_value_infos(implicit_args.items())
+    output_value_infos = [onnx.helper.make_value_info(name, TypeProto()) for name in outputs]
+
+    graph = onnx.helper.make_graph(
+        [node], "node_graph", input_value_infos + implicit_value_infos, output_value_infos
+    )
+    opset_id = onnx.helper.make_opsetid(schema.domain, schema.since_version)
+    model = onnx.helper.make_model(
+        graph,
+        opset_imports=[opset_id],
+        ir_version=irbuilder.select_ir_version(schema.since_version, domain=schema.domain),
+    )
+    model = onnx.shape_inference.infer_shapes(model)
+    # onnx.checker.check_model(model)
+    try:
+        sess = _cache_(model, ["CPUExecutionProvider"])
+    except (Fail, InvalidGraph, InvalidArgument) as e:
+        raise RuntimeError(
+            f"Unable to create onnxruntime InferenceSession "
+            f"with onnx model\n{utils.proto2text(model)}"
+        ) from e
+
+    session_run_input = {name: arg for name, arg in zip(inputs, args) if name != ""}
+    session_run_input.update(implicit_args)
+
+    try:
+        result = sess.run(None, session_run_input)
+    except (RuntimeError, Fail) as e:
+        raise RuntimeError(
+            f"Unable to execute model operator {schema.name!r} due to {e!r}"
+            f"\ninput types:\n"
+            f"{pprint.pformat({k: type(v) for k, v in zip(inputs, args)})}"
+            f"\nmodified input types:\n"
+            f"{pprint.pformat({k: type(v) for k, v in session_run_input.items()})}"
+            f"\ninputs:\n{pprint.pformat(session_run_input)}\n{model}"
+        ) from e
+
+    # Map ORT output values to the onnxscript representation-type.
+    return [ort_to_os_value(x) for x in result]
+
+
+def schema_id(schema):
+    return schema.name, schema.domain, schema.since_version
+
+
+class ORTEvaluator(Evaluator):
```

Comment:
If the plan is to have multiple evaluator, I would split this file into two: one for the base class, the other one with everything specific to onnxruntime.

Reply:
That's how I started, but I ended up simplifying things to avoid cyclic dependencies. I am going to leave it like this (at least until all cyclic dependencies are resolved for now).